userspace/libsinsp/examples/test.cpp

```diff
@@ -96,88 +109,153 @@ int main(int argc, char **argv)
-                    if(nullptr != fd_info && (fd_info->get_l4proto() != SCAP_L4_UNKNOWN && fd_info->get_l4proto() != SCAP_L4_NA))
-                    {
-                        cout << "[" << fd_info->tostring() << "]:";
-                    }
-                }
-                else
-                {
-                    cout << get_event_category(ev->get_category()) << "]:";
+                    cout << "[" << fd_info->tostring() << "]:";
                 }
+            }
+            else if(ev->get_category() == EC_IO_READ || ev->get_category() == EC_IO_WRITE)
+            {
+                cout << get_event_category(ev->get_category()) << "]:";
 
-                sinsp_threadinfo *p_thr = thread->get_parent_thread();
-                int64_t parent_pid;
-                if(nullptr != p_thr)
+                sinsp_fdinfo_t* fd_info = ev->get_fd_info();
+                if(nullptr != fd_info && (fd_info->get_l4proto() != SCAP_L4_UNKNOWN && fd_info->get_l4proto() != SCAP_L4_NA))
                 {
-                    parent_pid = p_thr->m_pid;
+                    cout << "[" << fd_info->tostring() << "]:";
                 }
+            }
+            else
+            {
+                cout << get_event_category(ev->get_category()) << "]:";
+            }
 
-                cout << "[PPID=" << parent_pid << "]:"
-                          << "[PID=" << thread->m_pid << "]:"
-                          << "[TYPE=" << get_event_type(ev->get_type()) << "]:"
-                          << "[EXE=" << thread->get_exepath() << "]:"
-                          << "[CMD=" << cmdline << "]"
-                          << endl;
+            sinsp_threadinfo *p_thr = thread->get_parent_thread();
+            int64_t parent_pid;
+            if(nullptr != p_thr)
+            {
+                parent_pid = p_thr->m_pid;
             }
+
+            cout << "[PPID=" << parent_pid << "]:"
+                        << "[PID=" << thread->m_pid << "]:"
+                        << "[TYPE=" << get_event_type(ev->get_type()) << "]:"
+                        << "[EXE=" << thread->get_exepath() << "]:"
+                        << "[CMD=" << cmdline << "]"
+                        << endl;
         }
-        else
+    }
+    else
+    {
+        cout << "[EVENT]:[" << get_event_category(ev->get_category()) << "]:"
+                    << ev->get_name() << endl;
+    }
+}
+
+#define PROCESS_DEFAULTS "*%evt.num %evt.time %evt.category %container.id %proc.ppid %proc.pid %evt.type %proc.exe %proc.cmdline %evt.args"
+void json_dump(sinsp& inspector)
+{
+    // Initialize JSON formatters
+    static sinsp_evt_formatter* default_formatter = new sinsp_evt_formatter(&inspector, DEFAULT_OUTPUT_STR);
+    static sinsp_evt_formatter* process_formatter = new sinsp_evt_formatter(&inspector, PROCESS_DEFAULTS);
+    static sinsp_evt_formatter* net_formatter = new sinsp_evt_formatter(&inspector, PROCESS_DEFAULTS " %fd.name");
```

Comment:
I'd put these 3 as globals, and i'd `delete` them at end of main; wdyt?

Reply:
Btw is there any reason why you heap allocated these formatters?

Reply:
I didn't want them to be allocated if the output selected is not JSON and I also didn't want them to be allocated and deleted on every call to `json_dump`, but it might make more sense to make them global and initialize them if the output selected is JSON.

Reply:
I just got your question, I could define them this way and they would be destroyed automatically before exiting the program:
```C++
    static sinsp_evt_formatter default_formatter(&inspector, DEFAULT_OUTPUT_STR);
    static sinsp_evt_formatter process_formatter(&inspector, PROCESS_DEFAULTS);
    static sinsp_evt_formatter net_formatter(&inspector, PROCESS_DEFAULTS " %fd.name");
```

Do you think this would be better than defining them in the heap?

Reply:
> but it might make more sense to make them global and initialize them if the output selected is JSON.

I think this is the best solution IMO :) 
This way we account for
> I didn't want them to be allocated if the output selected is not JSON

(i mean, it is not a big deal allocating them even when unused, but still i agree with you that the most polished solution is to allocating them only when json output is enabled).